### PR TITLE
all: use zstdframe where sensible

### PR DESCRIPTION
**Pull request by dsnet**
Use the zstdframe package where sensible instead of plumbing around our own zstd.Encoder just for stateless operations.

This causes logtail to have a dependency on zstd,
but that's arguably okay since zstd support is implicit to the protocol between a client and the logging service. Also, virtually every caller to logger.NewLogger was manually setting up a zstd.Encoder anyways,
meaning that zstd was functionally always a dependency.

Updates #cleanup
Updates tailscale/corp#18514

### CODE REVIEW

File: logtail/logtail.go [hunk 1]
```diff
@@ -62,7 +63,10 @@ type Config struct {
 	Stderr         io.Writer       // if set, logs are sent here instead of os.Stderr
 	StderrLevel    int             // max verbosity level to write to stderr; 0 means the non-verbose messages only
 	Buffer         Buffer          // temp storage, if nil a MemoryBuffer
-	NewZstdEncoder func() Encoder  // if set, used to compress logs for transmission
+	CompressLogs   bool            // whether to compress the log uploads
+
+	// Deprecated: Use CompressUploads instead.
+	NewZstdEncoder func() Encoder // if set, used to compress logs for transmission
```
Comment (bradfitz) [inline logtail/logtail.go:69]: sure, but feel free to also just remove and fix corp when we import this. that's what I'd do at least. fewer hops.

File: logtail/logtail.go [hunk 2]
```diff
@@ -364,8 +370,18 @@ func (l *Logger) uploading(ctx context.Context) {
...
-		if l.zstdEncoder != nil && len(body) > 256 {
-			zbody := l.zstdEncoder.EncodeAll(body, nil)
+		if (l.compressLogs || l.zstdEncoder != nil) && len(body) > 256 {
+			var zbody []byte
+			switch {
+			case l.zstdEncoder != nil:
+				zbody = l.zstdEncoder.EncodeAll(body, nil)
+			case l.lowMem:
+				zbody = zstdframe.AppendEncode(nil, body,
+					zstdframe.FastestCompression, zstdframe.LowMemory(true))
```
Comment (bradfitz) [inline logtail/logtail.go:380]: you always want LowMemory or only on mobile?

Comment (dsnet) [reply logtail/logtail.go:380]: I think it's reasonable to only do it on mobile.

File: logtail/logtail.go (hunk 1 above)
Comment (dsnet) [reply logtail/logtail.go:69]: That's my intent.

**Pull request merged**